Mathlib/NumberTheory/Padics/Harmonic.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2023 Koundinya Vajjha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Koundinya Vajjha
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.Data.Int.Log
+/-!
+
+The nth Harmonic number is not an integer. We formalize the proof using
+2-adic valuations. This proof is due to Kürschák.
+
+Reference:
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/padicharmonicsum.pdf
+
+-/
+
+open BigOperators
+
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, 1 / (i + 1)
+
+lemma harmonic_pos (n : ℕ) (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => div_pos zero_lt_one (by norm_cast; linarith))
+  (by rwa [Finset.nonempty_range_iff])
+
+lemma harmonic_singleton {n c : ℕ} (hc : c ∈ Finset.range n) :
+    harmonic n = 1 / (c + 1) + ∑ x in Finset.range n \ {c}, 1 / ((x : ℚ) + 1) := by
+  unfold harmonic
+  rwa [add_comm, Finset.sum_eq_sum_diff_singleton_add (i := c)]
+
+lemma finset_range_sdiff_singleton_nonempty {c n : ℕ} (hn : 2 ≤ n) :
+    Finset.Nonempty (Finset.range n \ {c}) := by
+  rw [Finset.sdiff_nonempty, Finset.subset_singleton_iff, Finset.range_eq_empty_iff, not_or]
+  refine' ⟨by linarith, fun Hnot => _⟩
+  suffices n = 1 by linarith
+  rw [← Finset.card_range n, ← Finset.card_singleton c, Hnot]
+
+lemma harmonic_singleton_ne_zero {c n : ℕ} (hn : 2 ≤ n) :
+    ∑ x in Finset.range n \ {c}, 1 / (x + 1 : ℚ) ≠ 0 := by
+  refine' ne_of_gt <| Finset.sum_pos (fun i _ => div_pos zero_lt_one _)
+    (finset_range_sdiff_singleton_nonempty hn)
+  norm_cast; simp only [add_pos_iff, or_true]
+
+/-- If a rational is not a 2-adic integer, it is not an integer. -/
+theorem not_int_of_not_padic_int {a : ℚ} (H : 1 < padicNorm 2 a) :
+    ¬ a.isInt := by
+  suffices : a.den ≠ 1; simpa [Rat.isInt]
+  by_cases h : a = 0
+  · subst h; contradiction
+  · unfold padicNorm at H
+    split_ifs at H; contradiction
+    suffices padicValRat 2 a < 0 by
+      unfold padicValRat at this
```

Comment:
You should think about whether this unfolding is due to missing API.

Reply:
perhaps! i'm not really sure. changing that to `dsimp [padicValRat]` or `simp only [padicValRat]` also works. presumably those are ok? i really just need to unfold the definition once in that subproof...

Reply:
But it seems like some of this definition unfolding is in order to use the formula for `padicNorm` or to prove that `padicValRat` is negative. Could those be separate lemmas?

Reply:
I am adding a 
```lean
lemma padicValRat_def (p : ℕ) (q : ℚ) :
  padicValRat p q = padicValInt p q.num - padicValNat p q.den
    := rfl
```